### PR TITLE
8283577: SSLEngine.unwrap on read-only input ByteBuffer

### PR DESCRIPTION
**Pull request by ascarpino**
Hi,

I need a review of this fix to allow a read-only 'src' buffer to be used with SSLEngine.unwrap(). A temporary read-write buffer is created in the SSLCipher operation when a read-only buffer is passed. If the 'src' is read-write, there is no effect on the current operation

The PR also includes a CSR for an API implementation note to the SSLEngine.unwrap. The 'src' buffer may be modified during the decryption operation. 'unwrap()' has had this behavior forever, so there is no compatibility issue with this note. Using the 'src' buffer for in-place decryption was a performance decision.

Tony

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issues
 * [JDK-8283577](https://bugs.openjdk.java.net/browse/JDK-8283577): SSLEngine.unwrap on read-only input ByteBuffer
 * [JDK-8285603](https://bugs.openjdk.java.net/browse/JDK-8285603): SSLEngine.unwrap modifies input ByteBuffer (**CSR**)

### Reviewers
 * [Bradford Wetmore](https://openjdk.java.net/census#wetmore) (@bradfordwetmore - **Reviewer**) ⚠️ Review applies to [0e654bfd](https://git.openjdk.java.net/jdk/pull/8462/files/0e654bfd3b32122c5f232658f343cf20b30e8223)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8462/head:pull/8462` \
`$ git checkout pull/8462`

Update a local copy of the PR: \
`$ git checkout pull/8462` \
`$ git pull https://git.openjdk.java.net/jdk pull/8462/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8462`

View PR using the GUI difftool: \
`$ git pr show -t 8462`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8462.diff">https://git.openjdk.java.net/jdk/pull/8462.diff</a>

</details>

### CODE REVIEW

**Comment by mcpowers** [comment]
I don't yet count as a reviewer, but your changes look good to me. You might want to let IntelliJ check for problems before you integrate.

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        in.clear();
+        receive(client, out, in);
+
+        System.out.println("done");
+
+        // Send bytes from the client and make sure the server receives the same
+        in.clear();
+        out.clear();
+        String testString = "ASDF";
+        in.put(testString.getBytes()).flip();
```
Comment (bradfordwetmore) [inline test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:172]: If you're going to convert back from UTF_8 later, you should probably convert using getBytes(UTF_8) here.

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        System.out.println("1: Server receive: " + testResult);
+        if (!testString.equalsIgnoreCase(testResult)) {
+            throw new Exception("unequal");
+        }
+
+        // Send bytes from the server and make sure the client receives the same
+        out.clear();
+        in.clear();
+        System.out.println("2: Server send: " + testString);
+        in.put(testString.getBytes()).flip();
```
Comment (bradfordwetmore) [inline test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:188]: Same

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        send(server, in, out);
```
Comment (bradfordwetmore) [inline test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:189]: Did you want to try asReadOnlyBuffer() here also?

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        in.clear();
+        receive(client, out, in);
```
Comment (bradfordwetmore) [inline test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:191]: And here?

**Review by bradfordwetmore — COMMENTED** [review]
Finished the SSLEngine/tests, starting on the SSLCipher.  Here's the notes so far.

**Review by bradfordwetmore — APPROVED** [review]
All of the comments observed are minor.  

Please consider some of the test changes, but otherwise looks good.

File: src/java.base/share/classes/javax/net/ssl/SSLEngine.java [hunk 5]
```diff
@@ -673,6 +673,9 @@ public abstract SSLEngineResult wrap(ByteBuffer [] srcs, int offset,
      * @throws  IllegalStateException if the client/server mode
      *          has not yet been set.
      * @see     #unwrap(ByteBuffer, ByteBuffer[], int, int)
+     *
+     * @implNote The data in {@code src} may be modified during the decryption
```
Comment (XueleiFan) [inline src/java.base/share/classes/javax/net/ssl/SSLEngine.java:677]: It looks like a note for the API users to me.  Is apiNote tag more appropriate here?

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
+/*
```
[the remark this answers is not in the document]
Comment (ascarpino) [reply test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:1]: I can move it.. I created it from another test which happen to be in that directory

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        ByteBuffer in = ByteBuffer.allocate(maxData);
+        ByteBuffer out = ByteBuffer.allocate(maxData);
+
+        HandshakeStatus statusClient, statusServer;
+        client.beginHandshake();
+        server.beginHandshake();
+
+        // Do TLS handshake
+        do {
+            statusClient = doHandshake(client, out, in);
```
[the remark this answers is not in the document]
Comment (ascarpino) [reply test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:157]: I got it from somewhere that I don't remember off hand.  I'm just trying to get through the handshake.

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+            statusServer = doHandshake(server, in, out);
+        } while (statusClient != HandshakeStatus.NOT_HANDSHAKING ||
+            statusServer != HandshakeStatus.NOT_HANDSHAKING);
+
+        // Read NST
```
[the remark this answers is not in the document]
Comment (ascarpino) [reply test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:162]: New Session Ticket

File: src/java.base/share/classes/javax/net/ssl/SSLEngine.java (hunk 5 above)
Comment (ascarpino) [reply src/java.base/share/classes/javax/net/ssl/SSLEngine.java:677]: The CSR and this code is changing, the note I was adding was already documented, but its existing wording should be more clear.

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java (hunk 1 above)
Comment (ascarpino) [reply test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:172]: setting the input to UTF8 isn't a concern.  The latter line to decode it changes it from using the ByteBuffer.toString() to the contents of the ByteBuffer in a String.

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java (hunk 2 above)
Comment (ascarpino) [reply test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:188]: Same :)

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java (hunk 3 above)
Comment (ascarpino) [reply test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:189]: Yeah, they should have been.  I must have undid it to test something

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java (hunk 4 above)
Comment (ascarpino) [reply test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:191]: Same

File: src/java.base/share/classes/javax/net/ssl/SSLEngine.java (hunk 5 above)
Comment (XueleiFan) [reply src/java.base/share/classes/javax/net/ssl/SSLEngine.java:677]: I like the new update as specified in the CSR.  I added my name as Reviewer for the CSR.

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (bradfordwetmore) [reply test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:1]: Ah.  If you can, please do.  Thanks.

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (bradfordwetmore) [reply test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:157]: It was just an observation, no need to change.

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java (hunk 1 above)
Comment (bradfordwetmore) [reply test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:172]: You could use the default charsets for encoding and decoding.  i.e. 

        in.clear();
        receive(server, out.asReadOnlyBuffer(), in);
        byte[] ba = new byte[in.remaining()];
        in.get(ba);
        String testResult = new String(ba);

File: test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (bradfordwetmore) [reply test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java:162]: Duh, of course!  Yay, TLAs!!!  (Three Letter Acronyms...)

Feel free to expand if you're so inclined.  ;)

**Comment by XueleiFan** [comment]
As the specification has been indicate that the input buffer could be updated, what do you think if closing the bug as "Not an issue"?  I was just wondering if it really worthy the effort to make the code more complicated.

**Comment by ascarpino** [comment]
There is too much grey area.  It says the src buffer maybe modified, which one could interpret it cannot be a read-only, but that would still need clarification to explicitly say "no read only buffers".  And other than these internal 'in-place' crypto reason, there is no API reason to not allow read-only buffers as input.
I did think about closing the CSR as the text was already there about the src buffer, even thought it was using a different term.  But I felt strongly enough that I wanted to prevent that confusion in the future.

**Comment by XueleiFan** [comment]
> There is too much grey area. It says the src buffer maybe modified, which one could interpret it cannot be a read-only, but that would still need clarification to explicitly say "no read only buffers". And other than these internal 'in-place' crypto reason, there is no API reason to not allow read-only buffers as input. I did think about closing the CSR as the text was already there about the src buffer, even thought it was using a different term. But I felt strongly enough that I wanted to prevent that confusion in the future.

I think it is good to make the clarification with a CSR.

As the spec says, "The inbound network buffer, {@code src}, may be modified", applications cannot assume that the inbound network buffer cannot be modified (read-only) any longer.  It does not grant that an application will work with read-only inbound buffers for another JSSE provider.  As a result, an application that want to support read-only buffers may also need to support non-read-only buffers.  As makes it more complicated in both application layers and the JSSE implementation layer.  It may be simple that applications and JSSE implementation codes only need to handle non-read-only buffers.

Just my $0.02.  I will let you and Brad for the final decision.

**Comment by ascarpino** [comment]
I think read-only has been allowed by the spec from the start, we just never allowed it to work properly.  I don't think this opens up possible complications in the code.

**Pull request closed without merging**